Overflow when deriving SignalPiece for recursive data structures

First of all, thank you for this project, I'm enjoying it a lot!

## Describe the Bug

When compiling a recursive struct the following error is created:

```rust
#[derive(Serialize, PartialEq, Clone, SignalPiece)]
pub struct MessagesA {
    pub messages: Vec<MessagesA>,
}

error[E0275]: overflow evaluating the requirement `MessagesA: SignalPiece`
```

The same happens for mutually recursive data structures:
```rust
#[derive(Serialize, PartialEq, Clone, SignalPiece)]
pub struct MessagesA {
    pub messages: Vec<MessagesB>,
}

#[derive(Serialize, PartialEq, Clone, SignalPiece)]
pub struct MessagesB {
    pub messages: Vec<MessagesA>,
}
```

And for enums:
```rust
#[derive(Serialize, PartialEq, Clone, SignalPiece)]
pub enum Messages {
    Messages(Box<Messages>),
    None,
}
```


## Expected Behavior

While I am not sure if there is a deeper limitation but these data structures should probably be possible to support, they are at least in Rust and Dart. It would be nice if they were supported by rinf as well.

## Additional Context

```
rustc 1.87.0 (17067e9ac 2025-05-09)
rinf = "8.4.0"
```

--- a/flutter_package/example/lib/main.dart
+++ b/flutter_package/example/lib/main.dart
@@ -74,10 +74,7 @@ class MyHomePage extends StatelessWidget {
           SampleNumberInput(
             letter: 'HELLO FROM DART!',
             dummyOne: 25,
-            dummyTwo: SampleSchema(
-              sampleFieldOne: true,
-              sampleFieldTwo: false,
-            ),
+            dummyTwo: SampleSchema(sampleFieldOne: true, sampleFieldTwo: false),
             dummyThree: [4, 5, 6],
           ).sendSignalToRust();
         },

--- a/flutter_package/example/native/hub/src/signals/complex_types.rs
+++ b/flutter_package/example/native/hub/src/signals/complex_types.rs
@@ -6,7 +6,9 @@ use rinf::{DartSignal, RustSignal, SignalPiece};
 use serde::{Deserialize, Serialize};
 use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
 
-#[derive(Serialize, Deserialize, PartialEq, DartSignal, RustSignal, Clone)]
+#[derive(
+  Serialize, Deserialize, PartialEq, DartSignal, RustSignal, SignalPiece, Clone,
+)]
 pub enum SerdeData {
   PrimitiveTypes(PrimitiveTypes),
   OtherTypes(Box<OtherTypes>),
@@ -19,8 +21,11 @@ pub enum SerdeData {
     f2: TupleStruct,
     f3: Struct,
   },
+  ListWithMutualRecursion(List),
+  TreeWithMutualRecursion(Tree),
   TupleArray([u32; 3]),
   UnitVector(Vec<()>),
+  SimpleList(SimpleList),
   CStyleEnum(CStyleEnum),
   ComplexMap(BTreeMap<([u32; 2], [u8; 4]), ()>),
   EmptyTupleVariant(),
@@ -78,6 +83,21 @@ pub struct Struct {
   pub x: u32,
   pub y: u64,
 }
+
+#[derive(Serialize, Deserialize, PartialEq, SignalPiece, Clone)]
+pub enum List {
+  Empty,
+  Node(Box<SerdeData>, Box<List>),
+}
+
+#[derive(Serialize, Deserialize, PartialEq, SignalPiece, Clone)]
+pub struct Tree {
+  pub value: Box<SerdeData>,
+  pub children: Vec<Tree>,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, SignalPiece, Clone)]
+pub struct SimpleList(pub Option<Box<SimpleList>>);
 
 #[derive(Serialize, Deserialize, PartialEq, SignalPiece, Clone)]
 pub enum CStyleEnum {

--- a/flutter_package/example/native/hub/src/testing/mod.rs
+++ b/flutter_package/example/native/hub/src/testing/mod.rs
@@ -1,7 +1,7 @@
 use crate::signals::{
-  CStyleEnum, ComplexSignalTestResult, NewTypeStruct, OtherTypes,
-  PrimitiveTypes, SerdeData, Struct, TupleStruct, UnitStruct, UnitTestEnd,
-  UnitTestStart,
+  CStyleEnum, ComplexSignalTestResult, List, NewTypeStruct, OtherTypes,
+  PrimitiveTypes, SerdeData, SimpleList, Struct, Tree, TupleStruct, UnitStruct,
+  UnitTestEnd, UnitTestStart,
 };
 use rinf::{DartSignal, RustSignal};
 use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
@@ -172,9 +172,51 @@ fn get_complex_signals() -> Vec<SerdeData> {
     f3: Struct { x: 4, y: 5 },
   };
 
+  let v7 = SerdeData::ListWithMutualRecursion(List::Empty);
+
+  let v8 = SerdeData::TreeWithMutualRecursion(Tree {
+    value: Box::new(SerdeData::PrimitiveTypes(PrimitiveTypes {
+      f_bool: false,
+      f_u8: 0,
+      f_u16: 1,
+      f_u32: 2,
+      f_u64: 3,
+      f_u128: 4,
+      f_i8: 5,
+      f_i16: 6,
+      f_i32: 7,
+      f_i64: 8,
+      f_i128: 9,
+      f_f32: None,
+      f_f64: None,
+      f_char: None,
+    })),
+    children: vec![Tree {
+      value: Box::new(SerdeData::PrimitiveTypes(PrimitiveTypes {
+        f_bool: false,
+        f_u8: 0,
+        f_u16: 0,
+        f_u32: 0,
+        f_u64: 0,
+        f_u128: 0,
+        f_i8: 0,
+        f_i16: 0,
+        f_i32: 0,
+        f_i64: 0,
+        f_i128: 0,
+        f_f32: None,
+        f_f64: None,
+        f_char: None,
+      })),
+      children: vec![],
+    }],
+  });
+
   let v9 = SerdeData::TupleArray([0, 2, 3]);
 
   let v10 = SerdeData::UnitVector(vec![(); 1000]);
+
+  let v11 = SerdeData::SimpleList(SimpleList(Some(Box::new(SimpleList(None)))));
 
   let v12 = SerdeData::CStyleEnum(CStyleEnum::C);
 
@@ -188,6 +230,7 @@ fn get_complex_signals() -> Vec<SerdeData> {
   let v15 = SerdeData::EmptyStructVariant {};
 
   vec![
-    v0, v1, v2, v2bis, v2ter, v3, v4, v5, v6, v9, v10, v12, v13, v14, v15,
+    v0, v1, v2, v2bis, v2ter, v3, v4, v5, v6, v7, v8, v9, v10, v11, v12, v13,
+    v14, v15,
   ]
 }

--- a/flutter_package/example/pubspec.yaml
+++ b/flutter_package/example/pubspec.yaml
@@ -5,7 +5,7 @@ publish_to: "none"
 version: 0.1.0
 
 environment:
-  sdk: ">=3.5.0 <4.0.0"
+  sdk: ">=3.8.0 <4.0.0"
 
 dependencies:
   flutter:

--- a/flutter_package/lib/rinf.dart
+++ b/flutter_package/lib/rinf.dart
@@ -47,9 +47,5 @@ void sendDartSignal(
   Uint8List messageBytes,
   Uint8List binary,
 ) {
-  sendDartSignalReal(
-    endpointSymbol,
-    messageBytes,
-    binary,
-  );
+  sendDartSignalReal(endpointSymbol, messageBytes, binary);
 }

--- a/flutter_package/lib/src/interface_os.dart
+++ b/flutter_package/lib/src/interface_os.dart
@@ -11,9 +11,7 @@ void setCompiledLibPathReal(String path) {
   overrideLibPath(path);
 }
 
-Future<void> prepareInterfaceReal(
-  AssignRustSignal assignRustSignal,
-) async {
+Future<void> prepareInterfaceReal(AssignRustSignal assignRustSignal) async {
   // Prepare ports for communication over isolates.
   final rustSignalPort = ReceivePort();
 

--- a/flutter_package/lib/src/interface_web.dart
+++ b/flutter_package/lib/src/interface_web.dart
@@ -13,22 +13,17 @@ void setCompiledLibPathReal(String path) {
   setJsLibPath(path);
 }
 
-Future<void> prepareInterfaceReal(
-  AssignRustSignal assignRustSignal,
-) async {
+Future<void> prepareInterfaceReal(AssignRustSignal assignRustSignal) async {
   // Load the JavaScript module.
   await loadJsFile();
 
   // Listen to Rust via JavaScript.
-  rinfBindingsObject['rinf_send_rust_signal_extern'] = ((
-    String endpoint,
-    JSObject messageBytesJS,
-    JSObject binaryJS,
-  ) {
-    final messageBytes = (messageBytesJS as JSUint8Array).toDart;
-    final binary = (binaryJS as JSUint8Array).toDart;
-    assignRustSignal[endpoint]!(messageBytes, binary);
-  }).toJS;
+  rinfBindingsObject['rinf_send_rust_signal_extern'] =
+      ((String endpoint, JSObject messageBytesJS, JSObject binaryJS) {
+        final messageBytes = (messageBytesJS as JSUint8Array).toDart;
+        final binary = (binaryJS as JSUint8Array).toDart;
+        assignRustSignal[endpoint]!(messageBytes, binary);
+      }).toJS;
 }
 
 void startRustLogicReal() {

--- a/flutter_package/lib/src/load_os.dart
+++ b/flutter_package/lib/src/load_os.dart
@@ -82,28 +82,17 @@ typedef PostCObjectInner = Int8 Function(Int64, Pointer<Dart_CObject>);
 typedef PostCObjectPtr = Pointer<NativeFunction<PostCObjectInner>>;
 typedef PrepareIsolateExtern = Void Function(PostCObjectPtr, Int64);
 typedef PrepareIsolateWrapped = void Function(PostCObjectPtr, int);
-typedef SendDartSignalExtern = Void Function(
-  Pointer<Uint8>,
-  UintPtr,
-  Pointer<Uint8>,
-  UintPtr,
-);
-typedef SendDartSignalWrapped = void Function(
-  Pointer<Uint8>,
-  int,
-  Pointer<Uint8>,
-  int,
-);
+typedef SendDartSignalExtern =
+    Void Function(Pointer<Uint8>, UintPtr, Pointer<Uint8>, UintPtr);
+typedef SendDartSignalWrapped =
+    void Function(Pointer<Uint8>, int, Pointer<Uint8>, int);
 
 /// Abstract class for unifying the interface
 /// for calling native functions.
 abstract class RustLibrary {
   void startRustLogic();
   void stopRustLogic();
-  void prepareIsolate(
-    PostCObjectPtr storePostObject,
-    int port,
-  );
+  void prepareIsolate(PostCObjectPtr storePostObject, int port);
   void sendDartSignal(
     String endpointSymbol,
     Uint8List messageBytes,
@@ -128,16 +117,10 @@ class RustLibraryGlobal extends RustLibrary {
 
   RustLibraryGlobal(this.lib);
 
-  @Native<Void Function()>(
-    isLeaf: true,
-    symbol: 'rinf_start_rust_logic_extern',
-  )
+  @Native<Void Function()>(isLeaf: true, symbol: 'rinf_start_rust_logic_extern')
   external static void startRustLogicExtern();
 
-  @Native<Void Function()>(
-    isLeaf: true,
-    symbol: 'rinf_stop_rust_logic_extern',
-  )
+  @Native<Void Function()>(isLeaf: true, symbol: 'rinf_stop_rust_logic_extern')
   external static void stopRustLogicExtern();
 
   @Native<PrepareIsolateExtern>(
@@ -186,10 +169,10 @@ class RustLibraryGlobal extends RustLibrary {
     // to reduce symbol lookup overhead.
     var sendDartSignalExtern = sendDartSignalExterns[endpointSymbol];
     if (sendDartSignalExtern == null) {
-      sendDartSignalExtern =
-          lib.lookupFunction<SendDartSignalExtern, SendDartSignalWrapped>(
-        endpointSymbol,
-      );
+      sendDartSignalExtern = lib
+          .lookupFunction<SendDartSignalExtern, SendDartSignalWrapped>(
+            endpointSymbol,
+          );
       sendDartSignalExterns[endpointSymbol] = sendDartSignalExtern;
     }
 
@@ -223,10 +206,10 @@ class RustLibraryLocal extends RustLibrary {
     stopRustLogicExtern = lib.lookupFunction<Void Function(), void Function()>(
       'rinf_stop_rust_logic_extern',
     );
-    prepareIsolateExtern =
-        lib.lookupFunction<PrepareIsolateExtern, PrepareIsolateWrapped>(
-      'rinf_prepare_isolate_extern',
-    );
+    prepareIsolateExtern = lib
+        .lookupFunction<PrepareIsolateExtern, PrepareIsolateWrapped>(
+          'rinf_prepare_isolate_extern',
+        );
   }
 
   @override
@@ -260,10 +243,10 @@ class RustLibraryLocal extends RustLibrary {
     // to reduce symbol lookup overhead.
     var sendDartSignalExtern = sendDartSignalExterns[endpointSymbol];
     if (sendDartSignalExtern == null) {
-      sendDartSignalExtern =
-          lib.lookupFunction<SendDartSignalExtern, SendDartSignalWrapped>(
-        endpointSymbol,
-      );
+      sendDartSignalExtern = lib
+          .lookupFunction<SendDartSignalExtern, SendDartSignalWrapped>(
+            endpointSymbol,
+          );
       sendDartSignalExterns[endpointSymbol] = sendDartSignalExtern;
     }
 

--- a/flutter_package/lib/src/load_web.dart
+++ b/flutter_package/lib/src/load_web.dart
@@ -48,14 +48,15 @@ Future<void> loadJsFile() async {
   // This will load the webassembly module.
   final scriptElement = HTMLScriptElement();
   scriptElement.type = 'module';
-  scriptElement.innerHTML = '''
+  scriptElement.innerHTML =
+      '''
 import init, * as wasmBindings from "$fullUrl";
 globalThis.wasmBindings = wasmBindings;
 await init();
 rinfBindings.completeRinfLoad();
 delete rinfBindings.completeRinfLoad;
 '''
-      .toJS;
+          .toJS;
   document.head!.append(scriptElement);
 
   // Await for the module to load.

--- a/flutter_package/pubspec.yaml
+++ b/flutter_package/pubspec.yaml
@@ -4,7 +4,7 @@ version: 8.4.0
 repository: https://github.com/cunarist/rinf
 
 environment:
-  sdk: ">=3.5.0 <4.0.0"
+  sdk: ">=3.8.0 <4.0.0"
   flutter: ">=3.22.0"
 
 dependencies:
@@ -14,7 +14,7 @@ dependencies:
   web: ^1.1.0
 
 dev_dependencies:
-  flutter_lints: ">=5.0.0 <7.0.0"
+  flutter_lints: ^6.0.0
 
 # For information on the generic Dart part of this file, see the
 # following page: https://dart.dev/tools/pub/pubspec

--- a/rust_crate/src/signal_trait.rs
+++ b/rust_crate/src/signal_trait.rs
@@ -40,7 +40,13 @@ pub trait DartSignalBinary: for<'a> Deserialize<'a> {
 }
 
 /// Enables a type to be nested within a signal struct or enum.
-pub trait SignalPiece {}
+pub trait SignalPiece {
+  /// This function is a no-op.
+  /// It's purely used for checking that
+  /// a field implements the `SignalPiece` trait.
+  #[doc(hidden)]
+  fn be_signal_piece(&self) {}
+}
 
 // Implement the trait for simple primitives.
 impl SignalPiece for i8 {}

--- a/rust_crate_proc/src/lib.rs
+++ b/rust_crate_proc/src/lib.rs
@@ -353,7 +353,7 @@ fn get_enum_signal_impl(
           .filter_map(|field| field.ident.clone())
           .collect();
         quote! {
-          Self::#variant_ident { #(ref #fields),* } => {
+          Self::#variant_ident { #(#fields),* } => {
             use rinf::SignalPiece;
             #(SignalPiece::be_signal_piece(#fields);)*
           }
@@ -369,7 +369,7 @@ fn get_enum_signal_impl(
           })
           .collect();
         quote! {
-          Self::#variant_ident(#(ref #field_vars),*) => {
+          Self::#variant_ident(#(#field_vars),*) => {
             use rinf::SignalPiece;
             #(SignalPiece::be_signal_piece(#field_vars);)*
           }

--- a/rust_crate_proc/src/lib.rs
+++ b/rust_crate_proc/src/lib.rs
@@ -2,7 +2,7 @@ use heck::{ToShoutySnakeCase, ToSnakeCase};
 use proc_macro::TokenStream;
 use quote::quote;
 use syn::{
-  Data, DataEnum, DataStruct, DeriveInput, Error, Fields, Ident,
+  Data, DataEnum, DataStruct, DeriveInput, Error, Fields, Ident, Index,
   parse_macro_input,
 };
 
@@ -30,15 +30,10 @@ pub fn derive_signal_piece(input: TokenStream) -> TokenStream {
   }
 
   // Enforce all fields to implement the foreign signal trait.
-  let where_clause = match &ast.data {
-    Data::Struct(data_struct) => get_struct_where_clause(data_struct),
-    Data::Enum(data_enum) => get_enum_where_clause(data_enum),
+  let expanded = match &ast.data {
+    Data::Struct(data_struct) => get_struct_signal_impl(data_struct, name),
+    Data::Enum(data_enum) => get_enum_signal_impl(data_enum, name),
     _ => return TokenStream::new(),
-  };
-
-  // Automatically implement the signal trait for the struct.
-  let expanded = quote! {
-    impl rinf::SignalPiece for #name #where_clause {}
   };
 
   // Convert the generated code into token stream and return it.
@@ -298,6 +293,103 @@ fn get_enum_where_clause(data_enum: &DataEnum) -> proc_macro2::TokenStream {
 
   quote! {
     where #(#variant_types: rinf::SignalPiece),*
+  }
+}
+
+fn get_struct_signal_impl(
+  data_struct: &DataStruct,
+  name: &Ident,
+) -> proc_macro2::TokenStream {
+  match &data_struct.fields {
+    Fields::Named(named_fields) => {
+      let fields = named_fields
+        .named
+        .iter()
+        .filter_map(|field| field.ident.clone());
+      quote! {
+        impl rinf::SignalPiece for #name {
+          fn be_signal_piece(&self) {
+            use rinf::SignalPiece;
+            #(SignalPiece::be_signal_piece(&self.#fields);)*
+          }
+        }
+      }
+    }
+    Fields::Unnamed(unnamed_fields) => {
+      let field_indices: Vec<Index> =
+        (0..unnamed_fields.unnamed.len()).map(Index::from).collect();
+      quote! {
+        impl rinf::SignalPiece for #name {
+          fn be_signal_piece(&self) {
+            use rinf::SignalPiece;
+            #(SignalPiece::be_signal_piece(&self.#field_indices);)*
+          }
+        }
+      }
+    }
+    Fields::Unit => {
+      quote! {
+        impl rinf::SignalPiece for #name {
+          fn be_signal_piece(&self) {
+            // Unit struct has no fields to check
+          }
+        }
+      }
+    }
+  }
+}
+
+fn get_enum_signal_impl(
+  data_enum: &DataEnum,
+  name: &Ident,
+) -> proc_macro2::TokenStream {
+  let variants = data_enum.variants.iter().map(|variant| {
+    let variant_ident = &variant.ident;
+    match &variant.fields {
+      Fields::Named(named_fields) => {
+        let fields: Vec<Ident> = named_fields
+          .named
+          .iter()
+          .filter_map(|field| field.ident.clone())
+          .collect();
+        quote! {
+          Self::#variant_ident { #(ref #fields),* } => {
+            use rinf::SignalPiece;
+            #(SignalPiece::be_signal_piece(#fields);)*
+          }
+        }
+      }
+      Fields::Unnamed(unnamed_fields) => {
+        let field_indices: Vec<Index> =
+          (0..unnamed_fields.unnamed.len()).map(Index::from).collect();
+        let field_vars: Vec<Ident> = field_indices
+          .iter()
+          .map(|i| {
+            Ident::new(&format!("field_{}", i.index), variant_ident.span())
+          })
+          .collect();
+        quote! {
+          Self::#variant_ident(#(ref #field_vars),*) => {
+            use rinf::SignalPiece;
+            #(SignalPiece::be_signal_piece(#field_vars);)*
+          }
+        }
+      }
+      Fields::Unit => {
+        quote! {
+          Self::#variant_ident => {}
+        }
+      }
+    }
+  });
+  quote! {
+    impl rinf::SignalPiece for #name {
+      fn be_signal_piece(&self) {
+        match self {
+          #( #variants )*
+        }
+      }
+    }
   }
 }
 

